File: cppCode/semantic.cpp
#include "semantic.h"
#include <iostream>

std::map<std::string, std::string> semanticAnalysis(ASTNode *ast, std::map<std::string, std::string> symbolTable)
{
    if (ast->type == "DECLARATION")
    {
        symbolTable[ast->value] = "int";
    }
    else if (ast->type == "ASSIGNMENT" || ast->type == "IDENTIFIER")
    {
        if (symbolTable.find(ast->value) == symbolTable.end())
        {
            throw std::runtime_error("Undefined variable: " + ast->value);
        }
    }
    for (ASTNode *child : ast->children)
    {
        symbolTable = semanticAnalysis(child, symbolTable);
    }
    std::cout << "Output from Semantic Analysis stage: (Symbol table not printed in C++, imagine JSON here)" << std::endl;
    return symbolTable;
}

File: JS/underConstruction/semantic.js
function semanticAnalysis(ast, symbolTable = {}) {
  if (ast.type === 'DECLARATION') {
    symbolTable[ast.value] = 'int';
  } else if (ast.type === 'ASSIGNMENT' || ast.type === 'IDENTIFIER') {
    if (!symbolTable[ast.value]) throw new Error(`Undefined variable: ${ast.value}`);
  }
  ast.children.forEach(child => semanticAnalysis(child, symbolTable));
  console.log('Output from Semantic Analysis stage:', JSON.stringify(symbolTable, null, 2));
  return symbolTable;
}

module.exports = semanticAnalysis;